server/src/test/java/org/elasticsearch/cluster/routing/allocation/IndexBalanceTests.java

```diff
@@ -440,4 +451,79 @@ public void testBalanceAllNodesStartedAddIndex() {
         assertThat(routingNodes.node("node2").shardsWithState("test1", STARTED).count(), equalTo(2L));
         assertThat(routingNodes.node("node3").shardsWithState("test1", STARTED).count(), equalTo(2L));
     }
+
+    /**
+     * {@see https://github.com/elastic/elasticsearch/issues/87279}
+     */
+    public void testRebalanceShouldNotPerformUnnecessaryMovesWithMultipleConcurrentRebalances() {
+        final var settings = Settings.builder()
+            .put("cluster.routing.allocation.cluster_concurrent_rebalance", randomIntBetween(3, 9))
+            .build();
+        final var allocationService = createAllocationService(settings);
+        assumeTrue(
+            "Only fixed in DesiredBalanceShardsAllocator",
+            allocationService.shardsAllocator instanceof DesiredBalanceShardsAllocator
+        );
```

Comment:
I think it would be nicer to change it so that `org.elasticsearch.cluster.ESAllocationTestCase#createShardsAllocator` respects the `cluster.routing.allocation.type` setting if set.